include/sparrow/array/array_data.hpp

```diff
@@ -52,6 +52,42 @@ namespace sparrow
         value_ptr<array_data> dictionary;
     };
 
+    data_descriptor type_descriptor(const array_data& data);
+    array_data::length_type length(const array_data& data);
+    std::int64_t offset(const array_data& data);
+
+    array_data::bitmap_type& bitmap(array_data& data);
+    const array_data::bitmap_type& bitmap(const array_data& data);
+
+    std::size_t buffers_size(const array_data& data);
+    array_data::buffer_type& buffer_at(array_data& data, std::size_t i);
+    const array_data::buffer_type& buffer_at(const array_data& data, std::size_t i);
+
+    std::size_t child_data_size(const array_data& data);
+    array_data& child_data_at(array_data& data, std::size_t i);
+    const array_data& child_data_at(const array_data& data, std::size_t i);
+
+    value_ptr<array_data>& dictionary(array_data& data);
+    const value_ptr<array_data>& dictionary(const array_data& data);
+
+    /**
+     * Concept for a structure that can be used as a data storage in the layout and the
+     * typed_array class.
+     */
+    template <class T>
+    concept data_storage = requires(const T t, std::size_t i)
+    {
+        { type_descriptor(t) } -> std::same_as<data_descriptor>;
+        { length(t) } -> std::same_as<std::int64_t>;
+        { offset(t) } -> std::same_as<std::int64_t>;
+        { bitmap(t) } -> std::ranges::random_access_range;
+        { buffers_size(t) } -> std::same_as<std::size_t>;
+        { buffer_at(t, i) } -> std::ranges::random_access_range;
+        { child_data_size(t) } -> std::same_as<std::size_t>;
+        child_data_at(t, i);
```

Comment:
```suggestion
        { child_data_at(t, i); } -> std::integer;
```
Would that be reasonable?

Reply:
actually `child_data_at` returns something that must satisfy the `data_storage` concept, so it's kind of recursive.